Algorithm.md

```diff
@@ -0,0 +1,30 @@
+1. Fetch available versions for each dependency
+2. If a project is already in the graph, intersect the new requirement with the one already present
+3. Select the highest [precedence] version which satisfies the intersected requirement (or only requirement, if this is the first insertion)
+  * This always has to be O(n) over the list of versions, because custom requirements mean that a previously failing version might become a candidate again
+4. Collect all child dependencies from that graph depth, repeat from step 1
+
+### In event of a conflict
+
+Given a project already in the graph with requirement X, and about to be re-added to the graph with requirement Y, where X and Y conflict:
+
+1. Find the _dependent_ project which changed the requirement into X (either from nil, or from a previously looser requirement)
+2. Select the next best version of that dependent project and attempt to finish resolving
```

Comment:
Can this be encoded as modifying the requirement, by intersecting with a more restrictive requirement which excludes the highest version previously selected?

That way we shouldn’t need a function that selects the “next highest” version from a list. There only needs to exist a way to, given a set of requirements, select the highest version and continue graph resolution.